Firebase/Messaging/FIRMessagingRmqManager.m

```diff
@@ -285,60 +315,34 @@ - (int64_t)queryLastRmqId {
 }
 
 - (NSArray *)unackedS2dRmqIds {
-  NSString *queryFormat = @"SELECT %@ FROM %@ ORDER BY %@ ASC";
-  NSString *query = [NSString stringWithFormat:queryFormat,
-                     kRmqIdColumn,
-                     kTableS2DRmqIds,
-                     kRmqIdColumn];
-  sqlite3_stmt *statement;
-  if (sqlite3_prepare_v2(_database, [query UTF8String], -1, &statement, NULL) != SQLITE_OK) {
-    FIRMessagingLoggerDebug(kFIRMessagingMessageCodeRmq2PersistentStore005,
-                            @"Could not find s2d ids");
-    _FIRMessagingRmqLogAndExit(statement, @[]);
-  }
-  NSMutableArray *rmqIDArray = [NSMutableArray array];
-  while (sqlite3_step(statement) == SQLITE_ROW) {
-    const char *rmqID = (char *)sqlite3_column_text(statement, 0);
-    [rmqIDArray addObject:[NSString stringWithUTF8String:rmqID]];
-  }
-  sqlite3_finalize(statement);
+  __block NSMutableArray *rmqIDArray = [NSMutableArray array];
+  dispatch_sync(_databaseOperationQueue, ^{
```

Comment:
Ideally, we should not use  `dispatch_sync` unless it is really necessary because it is pretty easy to loose track of all consequent calls and call e.g. `unackedS2dRmqIds` from `_databaseOperationQueue` which will lead to a deadlock.

Reply:
We def don't call a dispatch_sync call inside the queue, unackedS2dRmqIds is only called from outside. This class structure is quite simple that most of the writes are async and read is sync and can be ensure it's not called from a queue either.
Some of the read operations of database can not be async for now because either the public API is a sync call that calling this or we need to load data from database after app start right away because some other operation needs to access the data right away and a delay could result it thinks there's no such data. 

The most painful reason is that the original structure of this SDK is poorly designed for everything to be blocking call, so it will take a huge refactor to eventually move everything off main thread.
At this point it's still better to first move some of the operations off main thread, than everything on main thread.

Reply:
Yaeh, I agree and completely understand. Just would like to share our experience because we had issue with `dispatch_sync` just recently in GDT